src/stan/interface_callbacks/writer/base_writer.hpp

```diff
@@ -8,15 +9,55 @@ namespace stan {
   namespace interface_callbacks {
     namespace writer {
 
+      // base_writer is an absract base class defining the
+      // interface for Stan writer callbacks.  The Stan API
+      // writes structured output to implementations of this
+      // class defined by a given interface.
+      //
+      // A typical API function signature will be of the form
+      //
+      // template <class WriterA, class WriterB>
+      // void function(WriterA& writer_a, WriterB& writer_b) {
+      //   ...
+      //   writer_a(key, value);
+      //   writer_b("message");
+      //   ...
+      // }
+      //
+      // where WriterA and WriterB are base_writer implementations.
+
       class base_writer {
       public:
+        // Virtual destructor to avoid compiler warnings
         virtual ~base_writer() {}
-        // Can't enforce this method with a pure virtual function
-        // template <class T>
-        // virtual void operator()(const std::vector<T>& x) = 0;
-        virtual void operator()(const std::string x) = 0;
+
+        // Key-scalar input
+        virtual void operator()(const std::string& key,
```

Comment:
I don't see any reason it needs to be an operator from
a use point of view, but I'm OK with it.  I think of
operators like anonymous functions --- they apply
without the ".fun_name" bit.  How will they look in use?

Here's the comparison with an overloaded write function.
You construct a writer:

```
  writer_t w(...);
```

and then it's either the functor:

```
  w("foo");
```

or named method:

```
  w.write("foo");
```

It looks a bit different with longer variable names:

```
  writer_t writer(...);
```

and then

```
  writer("foo");
```

vs.

```
  writer.write("foo")
```


Also, do we have a write const char* that will work with "foo" without
forcing it to be constructed as a string (not without overhead --- there's
a malloc and copy of contents).  It could always be added as an optimization
later. 

On Oct 9, 2015, at 11:07 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/interface_callbacks/writer/base_writer.hpp:
> 
> >          virtual ~base_writer() {}
> > -        // Can't enforce this method with a pure virtual function
> > -        // template <class T>
> > -        // virtual void operator()(const std::vector<T>& x) = 0;
> > -        virtual void operator()(const std::string x) = 0;
> > +
> > +        // Key-scalar input
> > +        virtual void operator()(const std::string& key,
> 
> @bob-carpenter, think we should name these methods? The idea that this is a functor is nice, but I think the use here is a bit confusing. I don't have a suggestion for naming, though.
> 
> —
> Reply to this email directly or view it on GitHub.
>